File: packages/typescript/server/src/checkpoint.ts
/**
 * Checkpoint policy and scheduling for ANIP audit logs.
 */

import { createHash } from "crypto";
import type { StorageBackend } from "./storage.js";
import { MerkleTree } from "./merkle.js";
import { canonicalBytes } from "./hashing.js";

/**
 * Recursively sorted JSON serialization — matches Python's
 * json.dumps(obj, sort_keys=True, separators=(',', ':')).
 */
function canonicalJson(obj: unknown): string {
  if (obj === null || obj === undefined) return "null";
  if (typeof obj === "string") return JSON.stringify(obj);
  if (typeof obj === "number" || typeof obj === "boolean") return String(obj);
  if (Array.isArray(obj)) {
    return `[${obj.map(canonicalJson).join(",")}]`;
  }
  const keys = Object.keys(obj as Record<string, unknown>).sort();
  const pairs = keys.map(
    (k) =>
      `${JSON.stringify(k)}:${canonicalJson((obj as Record<string, unknown>)[k])}`,
  );
  return `{${pairs.join(",")}}`;
}

export interface CheckpointPolicyOpts {
  entryCount?: number;
  intervalSeconds?: number;
}

/**
 * Defines when a checkpoint should be created.
 */
export class CheckpointPolicy {
  private _entryCount: number | null;
  private _intervalSeconds: number | null;

  constructor(opts: CheckpointPolicyOpts) {
    this._entryCount = opts.entryCount ?? null;
    this._intervalSeconds = opts.intervalSeconds ?? null;
  }

  shouldCheckpoint(
    entriesSinceLast: number,
    secondsSinceLast: number = 0,
  ): boolean {
    if (this._entryCount !== null && entriesSinceLast >= this._entryCount) {
      return true;
    }
    if (
      this._intervalSeconds !== null &&
      secondsSinceLast >= this._intervalSeconds
    ) {
      return true;
    }
    return false;
  }
}

/**
 * Periodically triggers checkpoint creation using `setInterval`.
 *
 * The `createFn` callback is async — the caller is responsible for
 * leader acquisition and "has new entries" checks inside the callback.
 */
export class CheckpointScheduler {
  private _interval: number;
  private _createFn: () => Promise<void>;
  private _timer: ReturnType<typeof setInterval> | null = null;

  constructor(
    intervalSeconds: number,
    createFn: () => Promise<void>,
  ) {
    this._interval = intervalSeconds;
    this._createFn = createFn;
  }

  start(): void {
    this._timer = setInterval(async () => {
      try {
        await this._createFn();
      } catch {
        // Non-fatal
      }
    }, this._interval * 1000);
    if (this._timer && typeof this._timer === "object" && "unref" in this._timer) {
      (this._timer as NodeJS.Timeout).unref();
    }
  }

  stop(): void {
    if (this._timer !== null) {
      clearInterval(this._timer);
      this._timer = null;
    }
  }
}

export interface CreateCheckpointOpts {
  merkleSnapshot: { root: string; leaf_count: number };
  serviceId: string;
  previousCheckpoint: Record<string, unknown> | null;
  signFn?: (data: Buffer) => string;
}

/**
 * Create a checkpoint body and sign it.
 *
 * Returns `{ body, signature }`.
 */
export function createCheckpoint(opts: CreateCheckpointOpts): {
  body: Record<string, unknown>;
  signature: string;
} {
  const { merkleSnapshot, serviceId, previousCheckpoint, signFn } = opts;

  let firstSequence: number;
  let prevHash: string | null;
  let checkpointNumber: number;

  if (previousCheckpoint === null) {
    firstSequence = 1;
    prevHash = null;
    checkpointNumber = 1;
  } else {
    const prevRange = (previousCheckpoint.range as Record<string, number>) ?? {};
    firstSequence =
      (prevRange.last_sequence ?? (previousCheckpoint.last_sequence as number) ?? 0) + 1;

    const prevBodyCanonical = canonicalJson(previousCheckpoint);
    const hash = createHash("sha256")
      .update(prevBodyCanonical)
      .digest("hex");
    prevHash = `sha256:${hash}`;

    const prevId = (previousCheckpoint.checkpoint_id as string) ?? "ckpt-0";
    checkpointNumber = parseInt(prevId.split("-")[1], 10) + 1;
  }

  const lastSequence = merkleSnapshot.leaf_count;
  const entryCount = lastSequence - firstSequence + 1;

  const body: Record<string, unknown> = {
    version: "0.3",
    service_id: serviceId,
    checkpoint_id: `ckpt-${checkpointNumber}`,
    range: {
      first_sequence: firstSequence,
      last_sequence: lastSequence,
    },
    merkle_root: merkleSnapshot.root,
    previous_checkpoint: prevHash,
    timestamp: new Date().toISOString(),
    entry_count: entryCount,
  };

  const canonicalBodyBytes = Buffer.from(canonicalJson(body));
  const signature = signFn ? signFn(canonicalBodyBytes) : "";

  return { body, signature };
}

/**
 * Build a checkpoint by reading ALL audit entries from storage and
 * reconstructing the Merkle tree from scratch (cumulative, not delta).
 *
 * Returns `null` when there are no entries or no new entries since the
 * last stored checkpoint.
 */
export async function reconstructAndCreateCheckpoint(opts: {
  storage: StorageBackend;
  serviceId: string;
  signFn?: (data: Buffer) => string;
}): Promise<{ body: Record<string, unknown>; signature: string } | null> {
  const { storage, serviceId, signFn } = opts;

  const maxSeq = await storage.getMaxAuditSequence();
  if (maxSeq === null) return null;

  // Find the most-recent checkpoint to determine what is already covered.
  // getCheckpoints returns results in insertion order (oldest first),
  // so fetch all and take the last.
  const checkpoints = await storage.getCheckpoints(1_000_000);
  const lastCp =
    checkpoints.length > 0 ? checkpoints[checkpoints.length - 1] : null;
  const lastCovered = lastCp
    ? ((lastCp.range as Record<string, number>)?.last_sequence ??
        (lastCp.last_sequence as number) ??
        0)
    : 0;

  if (maxSeq <= lastCovered) return null;

  // Full reconstruction from entry 1
  const entries = await storage.getAuditEntriesRange(1, maxSeq);

  // Rebuild Merkle tree over every entry
  const tree = new MerkleTree();
  for (const entry of entries) {
    tree.addLeaf(Buffer.from(canonicalBytes(entry)));
  }

  const snapshot = tree.snapshot();

  // For cumulative checkpoints the range always starts at 1 because the
  // Merkle tree is rebuilt from the very first entry.  createCheckpoint
  // would normally compute firstSequence from the previous checkpoint's
  // last_sequence, so we pass a synthetic copy with last_sequence=0 to
  // force firstSequence=1 while keeping checkpoint_id for numbering.
  let syntheticPrev: Record<string, unknown> | null = null;
  if (lastCp !== null) {
    syntheticPrev = {
      ...lastCp,
      range: { ...(lastCp.range as Record<string, unknown>), last_sequence: 0 },
    };
  }

  const result = createCheckpoint({
    merkleSnapshot: snapshot,
    serviceId,
    previousCheckpoint: syntheticPrev,
    signFn,
  });

  // Restore the correct previous_checkpoint hash (computed from the real
  // stored checkpoint, not the synthetic copy).
  if (lastCp !== null) {
    const prevBodyCanonical = canonicalJson(lastCp);
    const hash = createHash("sha256").update(prevBodyCanonical).digest("hex");
    result.body.previous_checkpoint = `sha256:${hash}`;

    // Re-sign with the corrected body if a signFn was provided.
    if (signFn) {
      const correctedBytes = Buffer.from(canonicalJson(result.body));
      result.signature = signFn(correctedBytes);
    }
  }

  return result;
}

File: packages/go/server/checkpoint.go
package server

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"sort"
	"time"

	"github.com/anip-protocol/anip/packages/go/core"
	"github.com/anip-protocol/anip/packages/go/crypto"
)

// MerkleTree implements an RFC 6962 Merkle hash tree.
type MerkleTree struct {
	leaves [][]byte // leaf hashes (SHA256(0x00 || data))
}

// NewMerkleTree creates an empty Merkle tree.
func NewMerkleTree() *MerkleTree {
	return &MerkleTree{}
}

// AddLeaf adds a data element to the tree.
func (t *MerkleTree) AddLeaf(data []byte) {
	t.leaves = append(t.leaves, leafHash(data))
}

// LeafCount returns the number of leaves.
func (t *MerkleTree) LeafCount() int {
	return len(t.leaves)
}

// Root returns the Merkle root as "sha256:<hex>".
func (t *MerkleTree) Root() string {
	if len(t.leaves) == 0 {
		h := sha256.Sum256([]byte{})
		return fmt.Sprintf("sha256:%x", h)
	}
	root := t.computeRoot(0, len(t.leaves))
	return fmt.Sprintf("sha256:%x", root)
}

// InclusionProof generates the inclusion proof path for the leaf at index.
func (t *MerkleTree) InclusionProof(index int) ([]ProofStep, error) {
	if index < 0 || index >= len(t.leaves) {
		return nil, fmt.Errorf("leaf index %d out of range [0, %d)", index, len(t.leaves))
	}
	var path []ProofStep
	t.buildInclusionPath(index, 0, len(t.leaves), &path)
	return path, nil
}

// ConsistencyProof generates the consistency proof path from an older tree
// of size oldSize to this tree. The proof shows the old tree is a prefix
// of the current tree (RFC 6962 §2.1.4 consistency proof).
func (t *MerkleTree) ConsistencyProof(oldSize int) ([][]byte, error) {
	if oldSize < 0 || oldSize > len(t.leaves) {
		return nil, fmt.Errorf("old size %d out of range [0, %d]", oldSize, len(t.leaves))
	}
	if oldSize == 0 || oldSize == len(t.leaves) {
		return [][]byte{}, nil // Trivially consistent
	}
	return t.subproof(oldSize, 0, len(t.leaves), true), nil
}

// subproof implements RFC 6962 §2.1.4 SUBPROOF(m, D[lo:hi], start).
func (t *MerkleTree) subproof(m, lo, hi int, start bool) [][]byte {
	n := hi - lo
	if m == n {
		if !start {
			return [][]byte{t.computeRoot(lo, hi)}
		}
		return [][]byte{}
	}
	if m == 0 {
		return [][]byte{t.computeRoot(lo, hi)}
	}
	k := largestPowerOf2LessThan(n)
	if m <= k {
		path := t.subproof(m, lo, lo+k, start)
		path = append(path, t.computeRoot(lo+k, hi))
		return path
	}
	path := t.subproof(m-k, lo+k, hi, false)
	path = append(path, t.computeRoot(lo, lo+k))
	return path
}

// VerifyInclusion verifies that data at index is included in the tree.
func VerifyInclusion(data []byte, proof []ProofStep, expectedRoot string) bool {
	current := leafHash(data)
	for _, step := range proof {
		sibling := hexToBytes(step.Hash)
		if step.Side == "left" {
			current = nodeHash(sibling, current)
		} else {
			current = nodeHash(current, sibling)
		}
	}
	return fmt.Sprintf("sha256:%x", current) == expectedRoot
}

// ProofStep is a single step in a Merkle inclusion proof.
type ProofStep struct {
	Hash string `json:"hash"`
	Side string `json:"side"` // "left" or "right"
}

// leafHash computes SHA256(0x00 || data).
func leafHash(data []byte) []byte {
	input := append([]byte{core.LeafHashPrefix}, data...)
	h := sha256.Sum256(input)
	return h[:]
}

// nodeHash computes SHA256(0x01 || left || right).
func nodeHash(left, right []byte) []byte {
	input := append([]byte{core.NodeHashPrefix}, left...)
	input = append(input, right...)
	h := sha256.Sum256(input)
	return h[:]
}

// largestPowerOf2LessThan returns the largest power of 2 less than n.
func largestPowerOf2LessThan(n int) int {
	if n <= 1 {
		return 0
	}
	k := 1
	for k*2 < n {
		k *= 2
	}
	return k
}

// computeRoot recursively computes the Merkle root for leaves[lo:hi].
func (t *MerkleTree) computeRoot(lo, hi int) []byte {
	n := hi - lo
	if n == 1 {
		return t.leaves[lo]
	}
	split := largestPowerOf2LessThan(n)
	left := t.computeRoot(lo, lo+split)
	right := t.computeRoot(lo+split, hi)
	return nodeHash(left, right)
}

// buildInclusionPath recursively builds the inclusion proof path.
func (t *MerkleTree) buildInclusionPath(index, lo, hi int, path *[]ProofStep) {
	n := hi - lo
	if n == 1 {
		return
	}
	split := largestPowerOf2LessThan(n)
	if index-lo < split {
		t.buildInclusionPath(index, lo, lo+split, path)
		right := t.computeRoot(lo+split, hi)
		*path = append(*path, ProofStep{Hash: fmt.Sprintf("%x", right), Side: "right"})
	} else {
		t.buildInclusionPath(index, lo+split, hi, path)
		left := t.computeRoot(lo, lo+split)
		*path = append(*path, ProofStep{Hash: fmt.Sprintf("%x", left), Side: "left"})
	}
}

// hexToBytes converts a hex string to bytes.
func hexToBytes(h string) []byte {
	b := make([]byte, len(h)/2)
	for i := 0; i < len(h)/2; i++ {
		fmt.Sscanf(h[2*i:2*i+2], "%02x", &b[i])
	}
	return b
}

// canonicalBytes returns the canonical JSON bytes of an audit entry for Merkle leaf hashing.
// Matches the Python implementation: excludes "signature" and "id", sorts keys, compact separators.
func canonicalBytes(entry *core.AuditEntry) []byte {
	data, _ := json.Marshal(entry)
	var m map[string]any
	json.Unmarshal(data, &m)

	// Remove signature and id.
	delete(m, "signature")
	delete(m, "id")

	// Sort keys.
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	// Build with sorted keys using ordered map representation.
	ordered := make(map[string]any)
	for _, k := range keys {
		ordered[k] = m[k]
	}

	canonical, _ := json.Marshal(ordered)
	return canonical
}

// CreateCheckpoint builds a checkpoint from audit entries and stores it.
func CreateCheckpoint(
	km *crypto.KeyManager,
	storage Storage,
	serviceID string,
) (*core.Checkpoint, error) {
	// Get max sequence.
	maxSeq, err := storage.GetMaxAuditSequence()
	if err != nil {
		return nil, fmt.Errorf("get max audit sequence: %w", err)
	}
	if maxSeq == 0 {
		return nil, nil // No entries.
	}

	// Get the last checkpoint to determine the range.
	checkpoints, err := storage.ListCheckpoints(100)
	if err != nil {
		return nil, fmt.Errorf("list checkpoints: %w", err)
	}

	var lastCP *core.Checkpoint
	var lastCovered int
	if len(checkpoints) > 0 {
		lastCP = &checkpoints[len(checkpoints)-1]
		lastCovered = lastCP.Range["last_sequence"]
	}

	if maxSeq <= lastCovered {
		return nil, nil // No new entries.
	}

	// Full reconstruction from entry 1 (cumulative tree).
	entries, err := storage.GetAuditEntriesRange(1, maxSeq)
	if err != nil {
		return nil, fmt.Errorf("get audit entries range: %w", err)
	}

	// Build Merkle tree.
	tree := NewMerkleTree()
	for i := range entries {
		tree.AddLeaf(canonicalBytes(&entries[i]))
	}

	// Compute checkpoint number.
	cpNumber := 1
	var prevCheckpointHash string
	if lastCP != nil {
		// Parse number from last checkpoint ID.
		var n int
		if _, err := fmt.Sscanf(lastCP.CheckpointID, "ckpt-%d", &n); err == nil {
			cpNumber = n + 1
		}

		// Compute hash of previous checkpoint.
		prevBody, _ := json.Marshal(lastCP)
		// Re-serialize with sorted keys and compact separators.
		var prevMap map[string]any
		json.Unmarshal(prevBody, &prevMap)
		canonicalPrev, _ := json.Marshal(prevMap)
		h := sha256.Sum256(canonicalPrev)
		prevCheckpointHash = fmt.Sprintf("sha256:%x", h)
	}

	cp := &core.Checkpoint{
		Version:            "0.3",
		ServiceID:          serviceID,
		CheckpointID:       fmt.Sprintf("ckpt-%d", cpNumber),
		Range:              map[string]int{"first_sequence": 1, "last_sequence": maxSeq},
		MerkleRoot:         tree.Root(),
		PreviousCheckpoint: prevCheckpointHash,
		Timestamp:          time.Now().UTC().Format(time.RFC3339),
		EntryCount:         len(entries),
	}

	// Sign the checkpoint.
	cpJSON, _ := json.Marshal(cp)
	signature, err := crypto.SignDetachedJWSAudit(km, cpJSON)
	if err != nil {
		return nil, fmt.Errorf("sign checkpoint: %w", err)
	}

	// Store the checkpoint.
	if err := storage.StoreCheckpoint(cp, signature); err != nil {
		return nil, fmt.Errorf("store checkpoint: %w", err)
	}

	return cp, nil
}

// GenerateInclusionProof generates an inclusion proof for a leaf at the given index
// within the checkpoint's range. Returns the proof steps and the Merkle root.
// If entries have been deleted (expired), returns nil with a "proof_unavailable" indicator.
func GenerateInclusionProof(
	storage Storage,
	cp *core.Checkpoint,
	leafIndex int,
) ([]ProofStep, string, error) {
	firstSeq := cp.Range["first_sequence"]
	lastSeq := cp.Range["last_sequence"]

	// Get entries in the checkpoint range.
	entries, err := storage.GetAuditEntriesRange(firstSeq, lastSeq)
	if err != nil {
		return nil, "", fmt.Errorf("get audit entries: %w", err)
	}

	expectedCount := lastSeq - firstSeq + 1
	if len(entries) < expectedCount {
		// Entries have been deleted/expired.
		return nil, "audit_entries_expired", nil
	}

	// Rebuild Merkle tree.
	tree := NewMerkleTree()
	for i := range entries {
		tree.AddLeaf(canonicalBytes(&entries[i]))
	}

	// Validate leaf index.
	if leafIndex < 0 || leafIndex >= tree.LeafCount() {
		return nil, "", fmt.Errorf("leaf index %d out of range [0, %d)", leafIndex, tree.LeafCount())
	}

	proof, err := tree.InclusionProof(leafIndex)
	if err != nil {
		return nil, "", err
	}

	return proof, "", nil
}

// GenerateConsistencyProof generates a consistency proof between two checkpoints.
// It rebuilds the Merkle trees for both the old and new checkpoints and returns
// a proof that the old tree is a prefix of the new tree.
// If entries have been deleted (expired), returns nil with a "proof_unavailable" indicator.
func GenerateConsistencyProof(
	storage Storage,
	oldCP *core.Checkpoint,
	newCP *core.Checkpoint,
) (map[string]any, string, error) {
	oldLastSeq := oldCP.Range["last_sequence"]
	newLastSeq := newCP.Range["last_sequence"]

	// Rebuild old Merkle tree from entries 1..oldLastSeq.
	oldEntries, err := storage.GetAuditEntriesRange(1, oldLastSeq)
	if err != nil {
		return nil, "", fmt.Errorf("get old audit entries: %w", err)
	}
	if len(oldEntries) < oldLastSeq {
		return nil, "audit_entries_expired", nil
	}

	// Rebuild new Merkle tree from entries 1..newLastSeq.
	newEntries, err := storage.GetAuditEntriesRange(1, newLastSeq)
	if err != nil {
		return nil, "", fmt.Errorf("get new audit entries: %w", err)
	}
	if len(newEntries) < newLastSeq {
		return nil, "audit_entries_expired", nil
	}

	oldTree := NewMerkleTree()
	for i := range oldEntries {
		oldTree.AddLeaf(canonicalBytes(&oldEntries[i]))
	}

	newTree := NewMerkleTree()
	for i := range newEntries {
		newTree.AddLeaf(canonicalBytes(&newEntries[i]))
	}

	proof, err := newTree.ConsistencyProof(oldTree.LeafCount())
	if err != nil {
		return nil, "", fmt.Errorf("generate consistency proof: %w", err)
	}

	// Convert proof bytes to hex strings.
	path := make([]string, len(proof))
	for i, p := range proof {
		path[i] = fmt.Sprintf("%x", p)
	}

	result := map[string]any{
		"old_checkpoint_id": oldCP.CheckpointID,
		"new_checkpoint_id": newCP.CheckpointID,
		"old_size":          oldTree.LeafCount(),
		"new_size":          newTree.LeafCount(),
		"old_root":          oldTree.Root(),
		"new_root":          newTree.Root(),
		"path":              path,
	}

	return result, "", nil
}
